src/scaffold/services/helpers.tsx

```diff
@@ -0,0 +1,138 @@
+import React from "react";
+import { cloneDeep } from "lodash";
+import { GridEnrichedColDef } from "@mui/x-data-grid";
+
+export function hasValidChildren(name: string, children: any): boolean {
+  if (React.Children.count(children) === 0) return false;
+
+  if (React.Children.count(children) === 1) {
+    return children.type.name === name;
+  } else {
+    return children.some((child: JSX.Element) => child.type.name === name);
+  }
+}
+
+export function injectExtraColumns(
+  columns: GridEnrichedColDef[],
+  children: any,
+): GridEnrichedColDef[] {
+  const updatedColumns = cloneDeep(columns);
+
+  if (React.Children.count(children) === 1) {
+    if (children.type.name !== "ScaffoldExtraColumn") return updatedColumns;
+
+    const { props } = children;
+    const column: GridEnrichedColDef = {
+      field: props.label,
+      headerName: props.label,
+      flex: 1, // @todo flex
+    };
+
+    if (props.ValueComponent) {
+      const RenderCell = props.ValueComponent;
+      column.renderCell = ({ row }) => <RenderCell value={row} />;
+    }
+
+    updatedColumns.push(column);
+  } else {
+    children.forEach((child: JSX.Element, index: number) => {
+      if (child.type.name !== "ScaffoldExtraColumn") return;
+
+      const { props } = child;
+      // @todo flex
+      const column: GridEnrichedColDef = { field: props.field, flex: 1 };
+
+      if (props.label) {
+        column.headerName = props.label;
+      }
+
+      if (props.ValueComponent) {
+        const RenderCell = props.ValueComponent;
+        column.renderCell = ({ row }) => <RenderCell value={row} />;
+      }
+
+      updatedColumns.push(column);
+    });
+  }
+
+  return updatedColumns;
+}
+
+export function getColumnsFromChildren(children: any): GridEnrichedColDef[] {
```

Comment:
it's `any` here but this is also going to change a bit when refactoring for the design layer - started a chat with this in slack @christopherjbaker